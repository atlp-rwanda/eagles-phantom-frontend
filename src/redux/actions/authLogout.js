import axios from 'axios'
import * as actionTypes from '../types/logoutTypes';



export const logout = (history) => async(dispatch) => {
    try {
        
        dispatch(logoutRequest());
        const res = await axios.get('https://phanton-back-ft-logout--xiikam.herokuapp.com/api/v1/auth/logout')
        localStorage.removeItem('token')
        history.push("/login")
        dispatch(logoutSuccess(res))
        console.log('Response', res)
    } catch (error) {
        console.log(error)
        dispatch(logoutError(error))
    }    
};
    
export const logoutRequest = () =>{
    return { type: actionTypes.LOGOUT_REQUEST }
}  

export const logoutSuccess = (data) =>{
    
    return { 
        type: actionTypes.LOGOUT_SUCCESS,
        payload: data
        
    }
}  

export const logoutError = (error) =>{
    return { 
        type: actionTypes.LOGOUT_ERROR,
        payload: error
    }
}  




