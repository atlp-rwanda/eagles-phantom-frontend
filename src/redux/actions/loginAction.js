import { LOGIN_REQUEST, LOGIN_SUCCESS, LOGIN_FAIL } from "../types/loginTypes";
import axios from "axios";

export const loginAction = (data,history) => async (dispatch) => {

    try {
    dispatch(loginRequest());
    const res = await axios.post("https://eagles-phantom-backend.herokuapp.com/api/v1/auth/login", data);

    const user = await res.data;
    localStorage.setItem("token", user.token);
    localStorage.setItem('user-data', JSON.stringify(user));

    history.push('/dashboard')
    dispatch(loginSuccess(user));

    
    } catch (error) {
      if (error.response) {
        const errorMessage = await error.response.data.message;
        dispatch(loginFails(errorMessage))
    }
    else{
        dispatch(loginFails("Error, please check your connection and try again!"))
    }
    }
};

export const loginRequest = () => {
  return {
    type: LOGIN_REQUEST,
  };
};

export const loginSuccess = (data) => {
  return {
    type: LOGIN_SUCCESS,
    payload: data,
  };
};
export const loginFails = (error) => {
  return {
    type: LOGIN_FAIL,
    payload: error,
  };
};
