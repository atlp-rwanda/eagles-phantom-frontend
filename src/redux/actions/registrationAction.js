import React from 'react';
import {
  REGISTER,
  REGISTER_SUCCESS,
  REGISTER_FAIL,
} from "../types/registrationTypes";
import axios from "axios";

export const registrationAction = (data) => async (dispatch) => {
  try {
    dispatch(registration());
    const token = localStorage.getItem("token");
    const instance = axios.create({
      headers: { "x-access-token": token },
    });

    const res = await instance.post(
      "https://eagles-phantom-backend.herokuapp.com/api/v1/auth/register",
      data
    );
    

    // await axios.post(
    //   "http://localhost:3020/api/v1/auth/register",
    //   localStorage.setItem("x-access-token", token),
    //   data
    // );
    
    dispatch(registrationSuccess(res));
    // window.alert(res.data.message);
    // window.alert(res.data.message);
  } catch (error) {
    if (error) {
      // const errorMessage = error.response.data.message;
      // console.log(`this is the data ${data.message}`);
      // dispatch(registrationFail(errorMessage));
      dispatch(registrationFail(error));
      window.alert(error);
      // window.alert(errorMessage);
    } else {
      dispatch(registrationFail("Error in the Network"));
    }
  }
};

export const registration = () => {
  return {
    type: REGISTER,
  };
};

export const registrationSuccess = (data) => {
  return {
    type: REGISTER_SUCCESS,
    payload: data,
  };
};

export const registrationFail = (error) => {
  return {
    type: REGISTER_FAIL,
    payload: error,
  };
};
