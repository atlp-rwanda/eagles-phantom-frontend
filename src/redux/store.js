import {createStore, applyMiddleware } from 'redux';
import { composeWithDevTools} from 'redux-devtools-extension';
import thunk from 'redux-thunk';
import promise from 'redux-promise-middleware';
import allReducers from './reducers/';

// const rootReducer = combineReducers({
//   count: counterReducer,
// });

// const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const middleware = applyMiddleware(thunk, promise)
const store = createStore(allReducers, {}, composeWithDevTools(middleware));

export default store;
