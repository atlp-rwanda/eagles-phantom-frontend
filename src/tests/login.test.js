import React from "react";
import {
  render,
  cleanup,
  waitFor,
  waitForElementToBeRemoved,
} from '@testing-library/react';
import Login from "../components/loginPage/login";
import userEvent from "@testing-library/user-event";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import store from "../redux/store";
import reducer from "../redux/reducers/loginReducer";
import {
  LOGIN_REQUEST,
  LOGIN_SUCCESS,
  LOGIN_FAIL,
} from "../redux/types/loginTypes";
import mockAxios from "axios";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import { loginAction } from "../redux/actions/loginAction";

const onSubmit = jest.fn();

const LoginComponent = () => {
  return render(
    <Provider store={store}>
      <Router>
        <Login onSubmit={onSubmit}/>
      </Router>
    </Provider>
  );
};

const createMockStore = configureMockStore([thunk]);
const storeActions = createMockStore({
  login: {},
});

describe("LOGIN", () => {
  afterEach(cleanup);
  const initialState = {
    loading: "none",
    data: [],
    error: "",
  };
  const payload = {
    status: 200,
    message: "Login true",
    data: {
      id: 123,
      email: "email@gmail.com",
    },
  };

  it("should render the login component correctly", () => {
    const { asFragment } = LoginComponent();
    expect(asFragment(<Login />)).toMatchSnapshot();
  });

  it("should change state/value of email when typing", () => {
    const { getByLabelText } = LoginComponent();
    const email = getByLabelText("email");
    const password = getByLabelText("password");

    userEvent.type(email, "email@gmail.com");
    userEvent.type(password, "admin");

    expect(email.value).toBe("email@gmail.com");
    expect(password.value).toBe("admin");
  });

  it("should return error if password not provided", async () => {
    const { getByLabelText, container, getByText, debug } = LoginComponent();
    const email = getByLabelText("email");
    const form = container.querySelector("form");

    userEvent.type(email, "email@gmail.com");

    form.dispatchEvent(new Event("submit"));

    waitFor(() => expect(getByText("Provide a valid email")).toBeTruthy());
  });

  it("should return error if password not provided", async () => {
    const { getByLabelText, container, getByText } = LoginComponent();
    const password = getByLabelText("password");
    const form = container.querySelector("form");

    userEvent.type(password, "admin");

    form.dispatchEvent(new Event("submit"));

    waitFor(() => expect(getByText("Provide a password")).toBeTruthy());
  });

  it("should return error if password and email are not correct ", async () => {
    const { getByLabelText, container, getByText } = LoginComponent();
    const password = getByLabelText("password");
    const email = getByLabelText("email");
    const button = getByLabelText("login");

    userEvent.type(email, "email@gmail.com");
    userEvent.type(password, "password");

    button.click(button);

    waitFor(() =>
      expect(getByText("Invalid Username or password")).toBeTruthy()
    );
  });

  it("INITIAL STATE", () => {
    expect(reducer(undefined, {})).toEqual(initialState);
  });

  it("LOGIN_REQUEST ", () => {
    expect(reducer(initialState, { type: LOGIN_REQUEST })).toEqual({
      ...initialState,
      loading: "block",
    });
  });

  it("LOGIN_SUCCESS", () => {
    expect(
      reducer(initialState, {
        type: LOGIN_SUCCESS,
        payload,
      })
    ).toEqual({
      loading: "none",
      data: payload,
      error: "",
    });
  });

  it("LOGIN_FAIL", () => {
    const action = {
      type: LOGIN_FAIL,
      payload: "Login error",
    };

    expect(reducer(initialState, action)).toEqual({
      loading: "none",
      data: [],
      error: "Login error",
    });
  });
});
