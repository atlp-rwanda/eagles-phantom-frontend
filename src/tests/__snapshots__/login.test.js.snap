// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LOGIN should render the login component correctly 1`] = `
<DocumentFragment>
  <div
    class="login"
  >
    <form>
      <p
        id="loading"
        style="color: rgb(255, 255, 255); display: none;"
      >
        loading... 
      </p>
      <h1>
        Log in
      </h1>
      <span
        style="color: rgb(255, 255, 255);"
      >
         
        <br />
      </span>
      <input
        aria-label="email"
        id="email"
        name="email"
        placeholder="Email"
        type="email"
      />
      <br />
      <span
        style="color: rgb(255, 255, 255);"
      />
      <br />
       
      <br />
      <input
        aria-label="password"
        id="password"
        name="password"
        placeholder="Password"
        type="password"
      />
      <br />
      <span
        style="color: rgb(255, 255, 255);"
      />
      <br />
      <br />
      <input
        aria-label="login"
        class="login-btn"
        id="button"
        type="submit"
        value="LOGIN"
      />
      <br />
      <br />
      <a
        href=""
      >
        Forgot Password?
      </a>
      <br />
      <br />
    </form>
  </div>
</DocumentFragment>
`;
